File: python/add_two_numbers.py
class Solution:
    def addTwoNumbers(self, l1, l2):
        if not l1:
            return l2
        dummy = ListNode(-1)
        curr = dummy
        carry = 0
        while l1 or l2:
            res = carry
            res += l1.val if l1 else 0
            res += l2.val if l2 else 0

            carry = res / 10
            res = res % 10
            curr.next = ListNode(res)
            
            l1 = l1.next if l1 else None
            l2 = l2.next if l2 else None
            curr = curr.next
        curr.next = ListNode(carry) if carry else None
        return dummy.next

File: java/AddTwoNumbers.java
/**
 * Created by Rio on 1/4/15.
 * Add Two Numbers
 * https://oj.leetcode.com/problems/add-two-numbers/
 */
public class AddTwoNumbers {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        ListNode result = new ListNode(0);
        ListNode current = result;
        int carry = 0;

        while (l1 != null || l2 != null) {
            int sums = carry;
            sums += l1 != null ? l1.val : 0;
            sums += l2 != null ? l2.val : 0;

            carry = sums / 10;
            current.next = new ListNode(sums % 10);

            current = current.next;
            l1 = l1 == null ? null : l1.next;
            l2 = l2 == null ? null : l2.next;
        }
        current.next = carry == 0 ? null : new ListNode(carry);

        return result.next;
    }
}
